### DISCUSSION TITLE
ログイン時エラー文について

### DISCUSSION THREAD

**Discussion by nogogo** [pixiv/vroid-sdk-developers#61] [Q&A] [answered]
お世話になっております。

ログインに失敗した時、何故失敗したのかを分かりやすくユーザーに提示したいと考えております。
ApiErrorFormatクラス　
https://developer.vroid.com/sdk/docs/0.1.1/Pixiv.VroidSdk.Api.DataModel.ApiErrorFormat.html
でAPI関連のエラー文を取得できるかと思います。
ログイン処理失敗時のエラーの内、codeのパターンや、messageで受け取っているjsonの変数名"error"に対応する値のパターン、全て教えていただくことは可能でしょうか。

**Comment by nyamadan** [comment, marked as answer]
お世話になっております。

ログイン処理は[Login](https://developer.vroid.com/sdk/docs/0.1.1/Pixiv.VroidSdk.Oauth.Client.html#Pixiv_VroidSdk_Oauth_Client_Login_Pixiv_VroidSdk_Oauth_IBrowserControllable_System_Action_Pixiv_VroidSdk_Oauth_DataModel_Account__System_Action_System_Exception__)をExampleでは使用しております。
こちらはApiErrorFormatを使用しておりません。
[HttpRequestFailedException](https://developer.vroid.com/sdk/docs/0.1.1/Pixiv.VroidSdk.Networking.HttpRequestFailedException.html)をOnFailedで受け取った場合にログインにおける通信エラーが発生したとみなせます。


また、Exampleではログイン後[GetAccount](https://developer.vroid.com/sdk/docs/0.1.1/Pixiv.VroidSdk.Api.DefaultApi.html#Pixiv_VroidSdk_Api_DefaultApi_GetAccount_System_Action_Pixiv_VroidSdk_Api_DataModel_Account__System_Action_Pixiv_VroidSdk_Api_DataModel_ApiErrorFormat__)を呼び出しており、GetAccountが返す可能性があるエラーコード及びメッセージは次のようになっております。

 * "OAUTH_UNAUTHORIZED: OAuth unauthorized"
   * アプリ連携ができていない
 * "COMMON_SIGNED_IN_REQUIRED: signed in required"
   * ログインができていない
 * "OAUTH_FORBIDDEN: OAuth forbidden"
   * アプリの権限では実行できない

**Reply by nogogo** [reply]
ご回答ありがとうございます。
申し訳ありません。Legacyの方を使っているのを伝えておりませんでした。
Legacyの方ですと、認証コードをわざと間違えて連携しようとすると、エラーが出て、{"error":"invalid_request","error_description":"Missing required parameter: refresh_token."}
このようなjsonを.messageを見れば取得できるのですが、Legacyの方ですと、どのようなエラーが返ってきますでしょうか。

**Reply by nyamadan** [reply]
こちら、調査いたしました。

Legacyも同様にGetAccountをログイン後に実行しております。
調査した結果わかったことなのですが、API全般として以下の"error"が返ってくる可能性がございます。

 * "NETWORK_ERROR"
   * UnityWebRequestで[ネットワークエラー](https://docs.unity3d.com/ScriptReference/Networking.UnityWebRequest-isNetworkError.html)が発生した
 * "HTTP_ERROR"
   * UnityWebRequestで[HTTPエラー](https://docs.unity3d.com/ScriptReference/Networking.UnityWebRequest-isHttpError.html)が発生した
 * "AUTHORIZED_ERROR"
   * 認可失敗エラー、認可コードのコピー失敗やその他パラメータの設定ミス
 * "JSON_DESERIALIZE_ERROR"
   * レスポンスのJSONデシリアライズ失敗
  * "UNKNOWN_ERROR"
    * その他不明なエラー

**Reply by nogogo** [reply]
解決いたしました。
ご丁寧にありがとうございました。
